docs/design/2022-12-06-support-late-materialization.md

```diff
@@ -0,0 +1,144 @@
+
+There are two kinds of late materialization:
+1. Late materialization in selection
+
+    ![Fig.1](./imgs/late-materialization-in-selection.png)
+
+2. Late materialization in join （#Fig.2）
+
+    ![Fig.2](./imgs/late-materialization-in-join.png)
+
+**Only the late materialization in selection will be considered this time.**
+
+Late materialization can help improve performance of some AP queries. Like `SELECT * FROM t WHERE col_str LIKE '%PingCAP%'` The normal execution processes are as follows:
+- Read all the data pack of all the columns on table t
+- Run the filter `col_str LIKE '%PingCAP%'`
+- Return the result
+
+With late materialization, the execution processes can be:
+- Read all the data pack of col_str
+- Run the filter `col_str LIKE '%PingCAP%'`
+- Read the needed data pack of the rest columns on table t
+- Return the result
+
+And for more complex queries, like `SELECT * FROM t WHERE col_str1 LIKE '%PingCAP%' AND REGEXP_LIKE(col_str2, 'new\\*.\\*line')`, late materialization, the execution processes can be:
+- Read all the data pack of col_str
+- Run the filter `col_str1 LIKE '%PingCAP%'`
+- Read the needed data pack of rest columns on table t
+- Run the filter `REGEXP_LIKE(col_str2, 'new\\*.\\*line')`
+- Return the result
+
+Note that, `REGEXP_LIKE(col_str2, 'new\\*.\\*line')` is a heavy cost filter condition, filter out some data by `col_str LIKE '%PingCAP%'` first can reduce lots of useless calculation.
+
+Therefore, when late materialization is enabled, we can not only reduce the number of data packs to be read but also reduce some useless calculation. If the selectivity of the pushed down filter is considerable, the performance of the query can be highly improved.
+
+But TiFlash still does not support late materialization now. So, we are making a proposal to support it.
+
+## Objective
+- Improve performance of some AP queries.
+- Without performance degradation of most queries.
+
+## Design
+
+In brief, we will divide the filter conditions of selection(whose child is tablescan) into two parts: 
+- Those light and with high selectivity filter conditions, which will be pushed down to tablescan.
+- The rest filter conditions, which will be executed after tablescan.
+
+### TiDB side
+
+We will decide which filter conditions will be pushed down to tablescan in TiDB planner. We should consider the following factors:
+- Only the filter conditions with high selectivity should be pushed down.
+- The filter conditions which contain heavy cost functions should not be pushed down.
+- Filter conditions that apply to the same column are either pushed down or not pushed down at all.
+- The pushed down filter conditions should not contain too many columns.
+
+The algorithm is:
```

Comment:
It needs to be clearly stated in which optimizer phase this function is called. This is very important

Reply:
**postOptimize**

mentioned in line 96
```md
Since we need statistics to calculate the selectivity of the filter conditions, so this algorithm should be executed in postOptimize phase.
```